channels/serializers.py

```diff
@@ -37,3 +43,134 @@ def update(self, instance, validated_data):
+        return instance.url if not instance.is_self else None
+
+    def get_text(self, instance):
+        """Returns text or null depending on if it's a self post"""
+        return instance.selftext if instance.is_self else None
+
+    def get_upvoted(self, instance):
+        """Did the user upvote this?"""
+        return instance.likes is True
+
+    def get_downvoted(self, instance):
+        """Did the user downvote this?"""
+        return instance.likes is False
+
+    def get_created(self, instance):
+        """The ISO-8601 formatted datetime for the creation time"""
+        return datetime.fromtimestamp(instance.created, tz=timezone.utc).isoformat()
+
+    @staticmethod
+    def _parse_bool(value, field_name):
+        """Helper method to parse boolean values"""
+        if value in serializers.BooleanField.TRUE_VALUES:
+            return True
+        if value in serializers.BooleanField.FALSE_VALUES:
+            return False
+        raise ValidationError("{} must be a bool".format(field_name))
+
+    def validate_upvoted(self, value):
+        """Validate that upvoted is a bool"""
+        return {'upvoted': self._parse_bool(value, 'upvoted')}
+
+    def validate_downvoted(self, value):
+        """Validate that downvoted is a bool"""
+        return {'downvoted': self._parse_bool(value, 'downvoted')}
+
+    def validate_text(self, value):
+        """Validate that text is a string or null"""
+        if value is not None and not isinstance(value, str):
+            raise ValidationError("text must be a string")
+        return {"text": value}
+
+    def validate_url(self, value):
+        """Validate that URL is a string or null"""
+        if value is not None and not isinstance(value, str):
+            raise ValidationError("url must be a string")
+        return {"url": value}
+
+    def validate(self, attrs):
+        if attrs.get('downvoted') and attrs.get('upvoted'):
+            raise ValidationError("Only one of upvoted or downvoted can be true")
+
+        # url/text validation needs to go in created since it's handled differently
+        # than for update
+        return attrs
+
+    def create(self, validated_data):
+        title = validated_data['title']
+        text = validated_data.get('text')
+        url = validated_data.get('url')
+
+        if text and url:
+            raise ValidationError('Only one of text or url can be used to create a post')
+
+        if not text and not url:
+            raise ValidationError('One of text or url must be provided to create a post')
+
+        kwargs = {}
+        if text:
+            kwargs['text'] = text
+        else:
+            kwargs['url'] = url
+
+        api = Api(user=self.context['request'].user)
+        channel_name = self.context['view'].kwargs['channel_name']
+        return api.create_post(
+            channel_name,
+            title=title,
+            **kwargs
+        )
+
+    def update(self, instance, validated_data):
```

Comment:
Should it be an error to provide `True` for both `downvote` and `upvote` in the same request?